### PR TITLE
Immediately retry any requests that have backed off when a server comes back online.

### PR DESCRIPTION
**Pull request by erikjohnston**
Otherwise it can take up to a minute for any in-flight  `/send` requests to be retried.

### CODE REVIEW

**Review by DMRobertson — COMMENTED** [review]
Is there any danger that immediately retrying a bunch of requests ends up spamming the server with shedloads of stuff at once and brings it down again?

File: synapse/util/async_helpers.py [hunk 1]
```diff
@@ -766,3 +766,60 @@ def handle_cancel(new_deferred: "defer.Deferred[T]") -> None:
...
     return new_deferred
+
+
+class AwakenableSleeper:
+    """Allows explicitly waking up deferreds related to an entity that are
+    currently sleeping.
+    """
+
+    def __init__(self, reactor: IReactorTime) -> None:
+        self._streams: Dict[str, Set[defer.Deferred[None]]] = {}
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/async_helpers.py:777]: Oh I don't know, I have entirely lost the plot of when I can use what so I just do stuff and wait for the CI to tell me to fuck off :shrug:

File: synapse/util/async_helpers.py [hunk 2]
```diff
@@ -766,3 +766,60 @@ def handle_cancel(new_deferred: "defer.Deferred[T]") -> None:
...
+        self._reactor = reactor
+
+    def wake(self, name: str) -> None:
+        """Wake everything related to `name` that is currently sleeping."""
+        stream_set = self._streams.pop(name, set())
+        for deferred in set(stream_set):
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/async_helpers.py]: Ah, so, turns out this was due to another part of the code storing a `stream_set` across a yield point, and so popping from `self._streams` wasn't enough to ensure nothing else would try and mutate it. I think I've fixed that now though.

**Comment by erikjohnston** [comment]
> Is there any danger that immediately retrying a bunch of requests ends up spamming the server with shedloads of stuff at once and brings it down again?

I don't think so. We already clear the backoff timers when we see the host is back up, so this will only cause *in flight* requests to be immediately retried (and since we can only have one in flight transaction per remote host, that shouldn't be toooo bad)

**Review by DMRobertson — APPROVED** [review]
Many thanks. LGTM

File: synapse/util/retryutils.py [hunk 3]
```diff
@@ -239,6 +248,19 @@ async def store_retry_timings() -> None:
...
+                    self.notifier.notify_remote_server_up(self.destination)
+
+                if self.replication_client:
+                    # If we're on a worker we try and inform master about this. The
+                    # replication client doesn't hook into the notifier to avoid
+                    # infinite loops where we send a `REMOTE_SERVER_UP` command to
+                    # master, which then echoes it back to us which in turn pokes
+                    # the notifier.
+                    self.replication_client.send_remote_server_up(self.destination)
+
```
Comment (squahtx) [inline synapse/util/retryutils.py:263]: This path appears to get taken when we fail to contact the remote server too?

Comment (clokep) [reply synapse/util/retryutils.py:263]: See #16506.

**Pull request merged**